File: src/test/java/simplifying/conditional/expressions/consolidateConditionalExpression.java
package simplifying.conditional.expressions;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class consolidateConditionalExpression {
    private final AuthorizationService authorizationService = new AuthorizationService();
    private static final String ACTION = "An Action";

    @Test
    public void authorization_service_do_not_authorize_on_null_action() {
        assertThat(authorizationService.isAuthorized(null, null)).isFalse();
    }

    @Test
    public void authorization_service_do_not_authorize_on_empty_action() {
        assertThat(authorizationService.isAuthorized(null, "")).isFalse();
    }

    @Test
    public void authorization_service_do_not_authorize_disabled_user() {
        User disabledUser = User.builder().age(23).isDisabled(true).build();
        assertThat(authorizationService.isAuthorized(disabledUser, ACTION)).isFalse();
    }

    @Test
    public void authorization_service_do_not_authorize_not_loyal_user() {
        User disabledUser = User.builder().age(23).build();
        assertThat(authorizationService.isAuthorized(disabledUser, ACTION)).isFalse();
    }

    @Test
    public void authorization_service_do_not_authorize_on_age_less_than_12() {
        User tooYoungUser = User.builder().age(4).isLoyal(true).build();
        assertThat(authorizationService.isAuthorized(tooYoungUser, ACTION)).isFalse();
    }

    @Test
    public void authorization_service_authorize_loyal_major_enabled_user() {
        User authorizedUser = User.builder().age(23).isLoyal(true).build();
        assertThat(authorizationService.isAuthorized(authorizedUser, ACTION)).isTrue();
    }
}

File: src/scala/src/test/scala/simplifying/conditional/expressions/consolidateConditionalExpression.scala
package simplifying.conditional.expressions

import org.scalatest.flatspec.AnyFlatSpec

class consolidateConditionalExpression extends AnyFlatSpec {
  private val authorizationService = new AuthorizationService
  private val Action = "An Action"

  "Authorization service" should "not authorize when null action" in {
    assert(!authorizationService.isAuthorized(None, null))
  }

  "Authorization service" should "not authorize when None user" in {
    assert(!authorizationService.isAuthorized(None, Action))
  }

  "Authorization service" should "not authorize when empty action" in {
    assert(!authorizationService.isAuthorized(None, ""))
  }

  "Authorization service" should "not authorize when user is disabled" in {
    val disabledUser = Some(User(42, isDisabled = true, isLoyal = false))
    assert(!authorizationService.isAuthorized(disabledUser, Action))
  }

  "Authorization service" should "not authorize when user is enabled but not loyal" in {
    val notLoyalUser = Some(User(42, isDisabled = false, isLoyal = false))
    assert(!authorizationService.isAuthorized(notLoyalUser, Action))
  }

  "Authorization service" should "not authorize when user is minor" in {
    val minorUser = Some(User(17, isDisabled = false, isLoyal = true))
    assert(!authorizationService.isAuthorized(minorUser, Action))
  }

  "Authorization service" should "authorize when user is enabled, loyal and major" in {
    val loyalAndMajorUser = Some(User(42, isDisabled = false, isLoyal = true))
    assert(authorizationService.isAuthorized(loyalAndMajorUser, Action))
  }
}
